plugins/targetlocker/dblocker/dblocker.go

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package dblocker
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/facebookincubator/contest/pkg/logging"
+	"github.com/facebookincubator/contest/pkg/target"
+	"github.com/facebookincubator/contest/pkg/types"
+
+	// this blank import registers the mysql driver
+	_ "github.com/go-sql-driver/mysql"
+)
+
+// Name is the plugin name.
+var Name = "DBLocker"
+
+var log = logging.GetLogger("targetlocker/" + strings.ToLower(Name))
+
+// dblock represents parts of lock in the database, basically
+// a row from SELECT target_id, job_ID, expires_at
+type dblock struct {
+	targetId  string
+	jobID     int64
+	expiresAt time.Time
+}
+
+// String pretty-prints dblocks for logging and errors
+func (d dblock) String() string {
+	return fmt.Sprintf("target: %s, job %d, expires %s", d.targetId, d.jobID, d.expiresAt)
+}
+
+// targetIDList is a helper to convert contest targets to
+// a list of primary IDs used in the database
+func targetIDList(targets []*target.Target) []string {
+	res := make([]string, 0, len(targets))
+	for _, target := range targets {
+		res = append(res, target.ID)
+	}
+	return res
+}
+
+// listQueryString is a helper to create a (?, ?, ?) string
+// with as many ? as requested
+func listQueryString(length int) string {
+	switch length {
+	case 0:
+		return "()"
+	case 1:
+		return "(?)"
+	default:
+		return "(" + strings.Repeat("?, ", length-1) + "?)"
+	}
+}
+
+// DBLocker implements a simple target locker based on a relational database.
+// The current implementation only supports MySQL officially.
+// All functions in DBLocker are safe for concurrent by multiple goroutines.
+type DBLocker struct {
+	driverName string
+	db         *sql.DB
+	// lockTimeout set on each initial lock request
+	lockTimeout time.Duration
+	// refreshTimeout is used during refresh
+	refreshTimeout time.Duration
+}
+
+// cleanExpired deletes all expired locks on the given targets
+// (lock owner is ignored)
+func (d *DBLocker) cleanExpired(tx *sql.Tx, targets []string, now time.Time) error {
```

Comment:
`len(targets)` should be checked to be > 0

Reply:
this is marked as resolved - but why don't you check for zero length, instead of executing the DB query anyway?
```
if len(targets) == 0 {
    return nil
}
```

Reply:
I do, but right away in `Unlock`, `Lock`, and `RefreshLocks`, so we won't get here without locks :)

Reason I put it there is that there are 2 other places where it can create invalid queries with `... IN ()` (basically everywhere `listQueryString` is used)